### ISSUE TITLE
Generting vk.rs from the vk.xml

### ISSUE THREAD

**Issue by MaikKlein** [MaikKlein/ash#7]
Currently `vk.rs` was previously generated from https://github.com/Osspial/vk-rs, but it has been heavily modified since then.

It would be nice to generate it again to cut down on maintaining costs.

**Comment by asgeir** [comment]
Have you considered using the python scripts that are used for code generation for the [vulkan loader](https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/tree/master/scripts)?
If python is acceptable as a build-time dependency then this is something that is actively being used and is likely to be maintained. There is [another vulkan wrapper for rust](https://github.com/HellButcher/vulkan-rs/tree/master/tools/generator) that uses it and could be reused.

**Comment by MaikKlein** [comment]
@asgeir I have but I am not really familiar with python at all. Parsing isn't really a problem, I wanted to do it a few months ago with serde but the xml crate was heavily outdated. I sort of converted xml to json an tried to parse it but that caused even more problems.

But there is now https://crates.io/crates/serde-xml-rs and it looks usable.

**Comment by MaikKlein** [comment]
There is now https://github.com/terrybrashaw/vkxml which uses serde + a simpler version of vk.xml https://github.com/NicolBolas/New-Vulkan-XML-Format/tree/master.

The problem is with 1.1 https://github.com/NicolBolas/New-Vulkan-XML-Format/tree/master is currently broken.

There is also https://github.com/HellButcher/vulkan-rs/tree/master/generator but it is currently unmaintained and it parses the spec with xml.rs. It currently doesn't parse the 1.1 spec but it should be easy enough to fix.

Use serde-xml-rs to parse the vk.xml but I don't think that is possible because the spec is relatively hard to parse.

Write a parser for scratch with something like xml-rs or quick-xml.

Conclusion:
I think a serde parser seems to be the most easy to maintain but the vkxml crate is tied to New-Vulkan-XML-Format and I won't be able to maintain New-Vulkan-XML-Format myself.

I think I am going to try to parse the vk.xml with serde-xml-rs again, if that is not possible I am going to fork the vulkan-rs/generator.

**Comment by MaikKlein** [comment]
@krolli It doesn't need to be the same format as `vkxml::Registry`, I can always refactor. I also started to write an xml parser for the `vk.xml` which you can find here https://github.com/MaikKlein/ash/tree/generator/vk-registry. But I currently stopped at parsing function pointers, which are a bit annoying to parse to say the least.

How far along are you?

**Comment by krolli** [comment]
I'm currently doing rough pass on the whole thing, of which I miss just `<feature>` and `<extensions>` at the end. I wrote some quick hacks for pointers in structures and function parameters, but I will have to revisit them. First I wanted to have the whole thing roughly correct so I get the big picture.

I will probably stick to `vkxml::Registry` format for now since it makes it quite easy to find errors by comparing against reference, so refactoring shouldn't be needed.

**Comment by krolli** [comment]
So I got my parser to the point where it doesn't throw error on any vk.xml from v1.0.33 up to v1.1.75. It's not crates.io yet, but you can get it from https://github.com/krolli/vk-parse. I know of a few differences compared to output of ConvertToNewFmt.lua imported using vkxml:
- Preprocessor defines only have name imported, no body or list of parameters. This is next on my list of things to do.
- In v1.1 vk.xml has several elements and attributes added which are not currently exposed by vkxml crate or handled by ConvertToNewFmt.lua. Right now I discard them, though I'm considering creating my own structure to act as an IR between source xml and vkxml structures.
- I don't change whitespace in comments/notations.
- Sometimes I parse empty strings as None while original would give Some("").

I also seem to have stumbled upon some bugs in ConvertToNewFmt.lua which don't happen for me (AFAIK):
- sometimes hex constants are reported as c-expression
- sometimes extension type is discarded
